core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/config/update/PillarConfigUpdateService.java

```diff
@@ -0,0 +1,98 @@
+import java.util.List;
+import javax.inject.Inject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+@Component
+public class PillarConfigUpdateService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(PillarConfigUpdateService.class);
+
+    @Inject
+    private StackUpdater stackUpdater;
+
+    @Inject
+    private CloudbreakFlowMessageService flowMessageService;
+
+    @Inject
+    private ClusterService clusterService;
+
+    @Inject
+    private TransactionService transactionService;
+
+    @Inject
+    private ClusterCreationService clusterCreationService;
+
+    @Inject
+    private StackService stackService;
+
+    @Inject
+    private ClusterHostServiceRunner clusterHostServiceRunner;
+
+    public void doConfigUpdate(Long stackId) {
+        stackUpdater.updateStackStatus(stackId, DetailedStackStatus.BOOTSTRAPPING_MACHINES);
+        flowMessageService
+            .fireEventAndLog(stackId, UPDATE_IN_PROGRESS.name(),
+                CLUSTER_PILLAR_CONFIG_UPDATE_STARTED);
+        Stack stack = stackService.getByIdWithClusterInTransaction(stackId);
+        clusterHostServiceRunner.updateClusterConfigs(stack, stack.getCluster(), List.of());
+    }
+
+    public void configUpdateFinished(StackView stackView) {
+        try {
+            transactionService.required(() -> {
+                clusterService.updateClusterStatusByStackId(stackView.getId(), AVAILABLE);
+                stackUpdater.updateStackStatus(stackView.getId(), DetailedStackStatus.AVAILABLE,
+                    "Config update finished.");
+            });
+            flowMessageService.fireEventAndLog(stackView.getId(), AVAILABLE.name(),
+                CLUSTER_PILLAR_CONFIG_UPDATE_FINISHED);
+        } catch (TransactionExecutionException e) {
+            throw new TransactionRuntimeExecutionException(e);
+        }
+    }
+
+    public void handleConfigUpdateFailure(StackView stackView, Exception exception) {
+        try {
+            if (stackView.getClusterView() != null) {
+                String errorMessage = clusterCreationService
+                    .getErrorMessageFromException(exception);
+                transactionService.required(() -> {
+                    clusterService
+                        .updateClusterStatusByStackId(stackView.getId(), UPDATE_FAILED,
+                            errorMessage);
+                    stackUpdater
+                        .updateStackStatus(stackView.getId(), DetailedStackStatus.AVAILABLE);
+                });
+                flowMessageService.fireEventAndLog(stackView.getId(), UPDATE_FAILED.name(),
+                    CLUSTER_PILLAR_CONFIG_UPDATE_FAILED, errorMessage);
+            } else {
+                LOGGER.info("Cluster was null. Flow action was not required.");
+            }
+        } catch (TransactionExecutionException e) {
+            throw new TransactionRuntimeExecutionException(e);
```

Comment:
it's not ideal to throw any exception in the failure handling part

Reply:
Fixed